### PR TITLE
optionHome

### PR DESCRIPTION
**Pull request by lxc150896**

### CODE REVIEW

File: app/Http/ViewComposers/ProfileComposer.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        // $data['bannerBooks'] = Option::where(['key' => 'banner_book'])->first();
+        // $data['apps'] = Option::where(['key' => 'app'])->get();
+        // $data['textBanners'] = Option::where(['key' => 'text_banner'])->get();
+        // $data['textApps'] = Option::where(['key' => 'app_text'])->get();
+        // $data['contacts'] = Option::where(['key' => 'contact'])->get();
+        // $data['address'] = Option::where(['key' => 'address'])->get();
+        // $data['emails'] = Option::where(['key' => 'email'])->get();
+        // // dd($data);
+        
+        // view()->share($data);
```
Comment (framgiaciv3) [inline app/Http/ViewComposers/ProfileComposer.php]: **Reporter**: CHECKSTYLE</br>**Rule**: PSR2.Files.EndFileNewline.NoneFound</br>**Severity**: error</br>**File**: app/Http/ViewComposers/ProfileComposer.php</br>**Line**: 50</br>Expected 1 newline at end of file; 0 found

File: app/Http/Controllers/HomeController.php [hunk 2]
```diff
@@ -11,6 +11,7 @@
 use Auth;
 use Illuminate\Support\Facades\Session;
 use App\Http\Requests\SearchRequest;
```
Comment (framgiaciv3) [inline app/Http/Controllers/HomeController.php]: **Reporter**: CHECKSTYLE</br>**Rule**: PSR2.Namespaces.UseDeclaration.SpaceAfterLastUse</br>**Severity**: error</br>**File**: app/Http/Controllers/HomeController.php</br>**Line**: 13</br>There must be one blank line after the last USE statement; 0 found;

File: app/Providers/ComposerServiceProvider.php [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        $data['textFooters'] = Option::where(['key' => 'text_footer'])->get();
+        $data['banners'] = Option::where(['key' => 'banner'])->get();
+        $data['bannerBooks'] = Option::where(['key' => 'banner_book'])->first();
+        $data['apps'] = Option::where(['key' => 'app'])->get();
+        $data['textBanners'] = Option::where(['key' => 'text_banner'])->get();
+        $data['textApps'] = Option::where(['key' => 'app_text'])->get();
+        $data['contacts'] = Option::where(['key' => 'contact'])->get();
+        $data['address'] = Option::where(['key' => 'address'])->get();
+        $data['emails'] = Option::where(['key' => 'email'])->get();
+        return $data;
```
Comment (framgiaciv3) [inline app/Providers/ComposerServiceProvider.php]: **Reporter**: CHECKSTYLE</br>**Rule**: PSR2.Files.EndFileNewline.NoneFound</br>**Severity**: error</br>**File**: app/Providers/ComposerServiceProvider.php</br>**Line**: 49</br>Expected 1 newline at end of file; 0 found

File: app/Http/ViewComposers/ProfileComposer.php [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+class ProfileComposer
+{
+
+    /**
+     * Create a new profile composer.
+     *
+     * @return  void
+     */
+    protected $option;
```
Comment (framgiaciv3) [inline app/Http/ViewComposers/ProfileComposer.php]: **Reporter**: CHECKSTYLE</br>**Rule**: Squiz.Functions.FunctionDeclarationArgumentSpacing.SpacingBetween</br>**Severity**: error</br>**File**: app/Http/ViewComposers/ProfileComposer.php</br>**Line**: 19</br>Expected 0 spaces between parenthesis of function declaration; 0 found

File: app/Http/ViewComposers/ProfileComposer.php [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+    public function __construct(
```
Comment (framgiaciv3) [inline app/Http/ViewComposers/ProfileComposer.php]: **Reporter**: CHECKSTYLE</br>**Rule**: PSR2.Classes.PropertyDeclaration.Multiple</br>**Severity**: error</br>**File**: app/Http/ViewComposers/ProfileComposer.php</br>**Line**: 21</br>There must not be more than one property declared per statement

File: app/Http/ViewComposers/ProfileComposer.php [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    )
```
Comment (framgiaciv3) [inline app/Http/ViewComposers/ProfileComposer.php]: **Reporter**: CHECKSTYLE</br>**Rule**: Squiz.Functions.MultiLineFunctionDeclaration.NewlineBeforeOpenBrace</br>**Severity**: error</br>**File**: app/Http/ViewComposers/ProfileComposer.php</br>**Line**: 22</br>The closing parenthesis and the opening brace of a multi-line function declaration must be on the same line

File: app/Http/ViewComposers/ProfileComposer.php [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        // $data['bannerBooks'] = Option::where(['key' => 'banner_book'])->first();
+        // $data['apps'] = Option::where(['key' => 'app'])->get();
+        // $data['textBanners'] = Option::where(['key' => 'text_banner'])->get();
+        // $data['textApps'] = Option::where(['key' => 'app_text'])->get();
+        // $data['contacts'] = Option::where(['key' => 'contact'])->get();
+        // $data['address'] = Option::where(['key' => 'address'])->get();
+        // $data['emails'] = Option::where(['key' => 'email'])->get();
+        // dd($data);
+        
+        view()->with('option', $this->option->setting());
```
Comment (framgiaciv3) [inline app/Http/ViewComposers/ProfileComposer.php]: **Reporter**: CHECKSTYLE</br>**Rule**: PSR2.Files.EndFileNewline.NoneFound</br>**Severity**: error</br>**File**: app/Http/ViewComposers/ProfileComposer.php</br>**Line**: 50</br>Expected 1 newline at end of file; 0 found

File: app/Http/Controllers/HomeController.php (hunk 2 above)
Comment (framgiaciv3) [inline app/Http/Controllers/HomeController.php]: **Reporter**: CHECKSTYLE</br>**Rule**: PSR2.Namespaces.UseDeclaration.SpaceAfterLastUse</br>**Severity**: error</br>**File**: app/Http/Controllers/HomeController.php</br>**Line**: 13</br>There must be one blank line after the last USE statement; 0 found;

File: app/Providers/ComposerServiceProvider.php [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    //     $data['textFooters'] = Option::where(['key' => 'text_footer'])->get();
+    //     $data['banners'] = Option::where(['key' => 'banner'])->get();
+    //     $data['bannerBooks'] = Option::where(['key' => 'banner_book'])->first();
+    //     $data['apps'] = Option::where(['key' => 'app'])->get();
+    //     $data['textBanners'] = Option::where(['key' => 'text_banner'])->get();
+    //     $data['textApps'] = Option::where(['key' => 'app_text'])->get();
+    //     $data['contacts'] = Option::where(['key' => 'contact'])->get();
+    //     $data['address'] = Option::where(['key' => 'address'])->get();
+    //     $data['emails'] = Option::where(['key' => 'email'])->get();
+    //     return $data;
```
Comment (framgiaciv3) [inline app/Providers/ComposerServiceProvider.php]: **Reporter**: CHECKSTYLE</br>**Rule**: PSR2.Files.EndFileNewline.NoneFound</br>**Severity**: error</br>**File**: app/Providers/ComposerServiceProvider.php</br>**Line**: 49</br>Expected 1 newline at end of file; 0 found

File: app/Http/ViewComposers/ProfileComposer.php [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
```
Comment (framgiaciv3) [inline app/Http/ViewComposers/ProfileComposer.php]: **Reporter**: CHECKSTYLE</br>**Rule**: PSR2.Files.EndFileNewline.NoneFound</br>**Severity**: error</br>**File**: app/Http/ViewComposers/ProfileComposer.php</br>**Line**: 49</br>Expected 1 newline at end of file; 0 found

File: app/Http/Controllers/HomeController.php (hunk 2 above)
Comment (framgiaciv3) [inline app/Http/Controllers/HomeController.php]: **Reporter**: CHECKSTYLE</br>**Rule**: PSR2.Namespaces.UseDeclaration.SpaceAfterLastUse</br>**Severity**: error</br>**File**: app/Http/Controllers/HomeController.php</br>**Line**: 13</br>There must be one blank line after the last USE statement; 0 found;

File: app/Providers/ComposerServiceProvider.php (hunk 8 above)
Comment (framgiaciv3) [inline app/Providers/ComposerServiceProvider.php]: **Reporter**: CHECKSTYLE</br>**Rule**: PSR2.Files.EndFileNewline.NoneFound</br>**Severity**: error</br>**File**: app/Providers/ComposerServiceProvider.php</br>**Line**: 49</br>Expected 1 newline at end of file; 0 found

File: app/Http/Controllers/HomeController.php (hunk 2 above)
Comment (framgiaciv3) [inline app/Http/Controllers/HomeController.php]: **Reporter**: CHECKSTYLE</br>**Rule**: PSR2.Namespaces.UseDeclaration.SpaceAfterLastUse</br>**Severity**: error</br>**File**: app/Http/Controllers/HomeController.php</br>**Line**: 13</br>There must be one blank line after the last USE statement; 0 found;

File: app/Providers/ComposerServiceProvider.php (hunk 8 above)
Comment (framgiaciv3) [inline app/Providers/ComposerServiceProvider.php]: **Reporter**: CHECKSTYLE</br>**Rule**: PSR2.Files.EndFileNewline.NoneFound</br>**Severity**: error</br>**File**: app/Providers/ComposerServiceProvider.php</br>**Line**: 49</br>Expected 1 newline at end of file; 0 found

File: app/Helpers/helpers.php [hunk 10]
```diff
@@ -2,16 +2,20 @@
...
-        return null;
+    try {
+        $context = stream_context_create(['http' => ['header' => 'Connection: close\r\n']]);
+        $url = 'https://viblo.asia/';
+        $start = config('view.start');
+        $matchLeng = config('view.maxBanner');
+        $content = file_get_contents($url, false, $context, $start, $matchLeng);
+        $data = preg_match('/<div class="container px-0"><img src="(.*?)"/is', $content, $match);
+        if (isset($match[1])) {
+            return $match[1];
```
Comment (framgiaciv3) [inline app/Helpers/helpers.php]: **Reporter**: CHECKSTYLE</br>**Rule**: Generic.Files.LineLength.TooLong</br>**Severity**: warning</br>**File**: app/Helpers/helpers.php</br>**Line**: 15</br>Line exceeds 120 characters; contains 152 characters

File: app/Helpers/helpers.php [hunk 11]
```diff
@@ -26,20 +30,20 @@ function setTimeShort($value)
         $split = explode(' ', $time);
         switch ($split[1]) {
             case __('settings.time.full.seconds'):
-                return $split[0] . ' ' . __('settings.time.short.seconds');
-                break;
+            return $split[0] . ' ' . __('settings.time.short.seconds');
```
Comment (framgiaciv3) [inline app/Helpers/helpers.php]: **Reporter**: CHECKSTYLE</br>**Rule**: PSR2.ControlStructures.SwitchDeclaration.BreakIndent</br>**Severity**: error</br>**File**: app/Helpers/helpers.php</br>**Line**: 33</br>Terminating statement must be indented to the same level as the CASE body

File: app/Helpers/helpers.php [hunk 12]
```diff
@@ -26,20 +30,20 @@ function setTimeShort($value)
...
+            break;
             case __('settings.time.full.minutes'):
-                return $split[0] . ' ' . __('settings.time.short.minutes');
-                break;
+            return $split[0] . ' ' . __('settings.time.short.minutes');
```
Comment (framgiaciv3) [inline app/Helpers/helpers.php]: **Reporter**: CHECKSTYLE</br>**Rule**: PSR2.ControlStructures.SwitchDeclaration.BreakIndent</br>**Severity**: error</br>**File**: app/Helpers/helpers.php</br>**Line**: 36</br>Terminating statement must be indented to the same level as the CASE body

File: app/Helpers/helpers.php [hunk 13]
```diff
@@ -26,20 +30,20 @@ function setTimeShort($value)
...
+            break;
             case __('settings.time.full.hours'):
-                return $split[0] . ' ' . __('settings.time.short.hours');
-                break;
+            return $split[0] . ' ' . __('settings.time.short.hours');
```
Comment (framgiaciv3) [inline app/Helpers/helpers.php]: **Reporter**: CHECKSTYLE</br>**Rule**: PSR2.ControlStructures.SwitchDeclaration.BreakIndent</br>**Severity**: error</br>**File**: app/Helpers/helpers.php</br>**Line**: 39</br>Terminating statement must be indented to the same level as the CASE body

File: app/Helpers/helpers.php [hunk 14]
```diff
@@ -26,20 +30,20 @@ function setTimeShort($value)
...
+            break;
             case __('settings.time.full.days'):
-                return $split[0] . ' ' . __('settings.time.short.days');
-                break;
+            return $split[0] . ' ' . __('settings.time.short.days');
```
Comment (framgiaciv3) [inline app/Helpers/helpers.php]: **Reporter**: CHECKSTYLE</br>**Rule**: PSR2.ControlStructures.SwitchDeclaration.BreakIndent</br>**Severity**: error</br>**File**: app/Helpers/helpers.php</br>**Line**: 42</br>Terminating statement must be indented to the same level as the CASE body

File: app/Helpers/helpers.php [hunk 15]
```diff
@@ -26,20 +30,20 @@ function setTimeShort($value)
...
+            break;
             default:
-                return $split[0] . ' ' . __('settings.time.short.weeks');
-                break;
+            return $split[0] . ' ' . __('settings.time.short.weeks');
```
Comment (framgiaciv3) [inline app/Helpers/helpers.php]: **Reporter**: CHECKSTYLE</br>**Rule**: PSR2.ControlStructures.SwitchDeclaration.BreakIndent</br>**Severity**: error</br>**File**: app/Helpers/helpers.php</br>**Line**: 45</br>Terminating statement must be indented to the same level as the CASE body

File: app/Helpers/helpers.php [hunk 16]
```diff
@@ -67,28 +71,28 @@ function translate($value)
 {
     switch ($value) {
         case 'waiting':
-            return __('settings.book.waiting');
-            break;
+        return __('settings.book.waiting');
```
Comment (framgiaciv3) [inline app/Helpers/helpers.php]: **Reporter**: CHECKSTYLE</br>**Rule**: PSR2.ControlStructures.SwitchDeclaration.BreakIndent</br>**Severity**: error</br>**File**: app/Helpers/helpers.php</br>**Line**: 74</br>Terminating statement must be indented to the same level as the CASE body

File: app/Helpers/helpers.php [hunk 17]
```diff
@@ -67,28 +71,28 @@ function translate($value)
...
+        break;
         case 'reading':
-            return __('settings.book.reading');
-            break;
+        return __('settings.book.reading');
```
Comment (framgiaciv3) [inline app/Helpers/helpers.php]: **Reporter**: CHECKSTYLE</br>**Rule**: PSR2.ControlStructures.SwitchDeclaration.BreakIndent</br>**Severity**: error</br>**File**: app/Helpers/helpers.php</br>**Line**: 77</br>Terminating statement must be indented to the same level as the CASE body

File: app/Helpers/helpers.php [hunk 18]
```diff
@@ -67,28 +71,28 @@ function translate($value)
...
+        break;
         case 'returning':
-            return __('settings.book.returning');
-            break;
+        return __('settings.book.returning');
```
Comment (framgiaciv3) [inline app/Helpers/helpers.php]: **Reporter**: CHECKSTYLE</br>**Rule**: PSR2.ControlStructures.SwitchDeclaration.BreakIndent</br>**Severity**: error</br>**File**: app/Helpers/helpers.php</br>**Line**: 80</br>Terminating statement must be indented to the same level as the CASE body

File: app/Helpers/helpers.php [hunk 19]
```diff
@@ -67,28 +71,28 @@ function translate($value)
...
+        break;
         case 'returned':
-            return __('settings.book.returned');
-            break;
+        return __('settings.book.returned');
```
Comment (framgiaciv3) [inline app/Helpers/helpers.php]: **Reporter**: CHECKSTYLE</br>**Rule**: PSR2.ControlStructures.SwitchDeclaration.BreakIndent</br>**Severity**: error</br>**File**: app/Helpers/helpers.php</br>**Line**: 83</br>Terminating statement must be indented to the same level as the CASE body

File: app/Helpers/helpers.php [hunk 20]
```diff
@@ -67,28 +71,28 @@ function translate($value)
...
+        break;
         case config('view.notifications.waiting'):
-            return __('settings.notifications.waiting');
-            break;
+        return __('settings.notifications.waiting');
```
Comment (framgiaciv3) [inline app/Helpers/helpers.php]: **Reporter**: CHECKSTYLE</br>**Rule**: PSR2.ControlStructures.SwitchDeclaration.BreakIndent</br>**Severity**: error</br>**File**: app/Helpers/helpers.php</br>**Line**: 86</br>Terminating statement must be indented to the same level as the CASE body

File: app/Helpers/helpers.php [hunk 21]
```diff
@@ -67,28 +71,28 @@ function translate($value)
...
+        break;
         case config('view.notifications.returning'):
-            return __('settings.notifications.returning');
-            break;
+        return __('settings.notifications.returning');
```
Comment (framgiaciv3) [inline app/Helpers/helpers.php]: **Reporter**: CHECKSTYLE</br>**Rule**: PSR2.ControlStructures.SwitchDeclaration.BreakIndent</br>**Severity**: error</br>**File**: app/Helpers/helpers.php</br>**Line**: 89</br>Terminating statement must be indented to the same level as the CASE body

File: app/Helpers/helpers.php [hunk 22]
```diff
@@ -67,28 +71,28 @@ function translate($value)
...
+        break;
         case config('view.notifications.returned'):
-            return __('settings.notifications.returned');
-            break;
+        return __('settings.notifications.returned');
```
Comment (framgiaciv3) [inline app/Helpers/helpers.php]: **Reporter**: CHECKSTYLE</br>**Rule**: PSR2.ControlStructures.SwitchDeclaration.BreakIndent</br>**Severity**: error</br>**File**: app/Helpers/helpers.php</br>**Line**: 92</br>Terminating statement must be indented to the same level as the CASE body

File: app/Helpers/helpers.php [hunk 23]
```diff
@@ -67,28 +71,28 @@ function translate($value)
...
+        break;
         default:
-            return __('settings.notifications.reading');
-            break;
+        return __('settings.notifications.reading');
```
Comment (framgiaciv3) [inline app/Helpers/helpers.php]: **Reporter**: CHECKSTYLE</br>**Rule**: PSR2.ControlStructures.SwitchDeclaration.BreakIndent</br>**Severity**: error</br>**File**: app/Helpers/helpers.php</br>**Line**: 95</br>Terminating statement must be indented to the same level as the CASE body

File: app/Helpers/helpers.php [hunk 24]
```diff
@@ -2,17 +2,22 @@
...
-        return null;
+    try {
+        $context = stream_context_create(['http' => ['header' => 'Connection: close\r\n']]);
+        $url = 'https://viblo.asia/';
+        $start = config('view.start');
+        $matchLeng = config('view.maxBanner');
+        $content = file_get_contents($url, false, $context, $start, $matchLeng);
+        $data = preg_match('/<div class="container px-0"><img src="(.*?)"/is', $content, $match);
+        if (isset($match[1])) {
+            return $match[1];
```
Comment (framgiaciv3) [inline app/Helpers/helpers.php]: **Reporter**: CHECKSTYLE</br>**Rule**: Generic.Files.LineLength.TooLong</br>**Severity**: warning</br>**File**: app/Helpers/helpers.php</br>**Line**: 15</br>Line exceeds 120 characters; contains 152 characters

File: app/Helpers/helpers.php [hunk 25]
```diff
@@ -2,17 +2,22 @@
...
+        } else {
+            return $match[1] = 'https://viblo.asia/announcements/gioi-thieu-state-of-chain-tieu-diem-hackathon-meet-up-blockchain-nam-2018-YWOZrpmp5Q0';
+        }
+    } catch (Exception $e) {
+        return view('error');
     }
```
Comment (framgiaciv3) [inline app/Helpers/helpers.php]: **Reporter**: CHECKSTYLE</br>**Rule**: PSR2.Methods.FunctionClosingBrace.SpacingBeforeClose</br>**Severity**: error</br>**File**: app/Helpers/helpers.php</br>**Line**: 21</br>Function closing brace must go on the next line following the body; found 1 blank lines before brace

File: app/Helpers/helpers.php [hunk 26]
```diff
@@ -2,17 +2,23 @@
...
-        return null;
+    try {
+        $context = stream_context_create(['http' => ['header' => 'Connection: close\r\n']]);
+        $url = 'https://viblo.asia/';
+        $start = config('view.start');
+        $matchLeng = config('view.maxBanner');
+        $content = file_get_contents($url, false, $context, $start, $matchLeng);
+        $data = preg_match('/<div class="container px-0"><img src="(.*?)"/is', $content, $match);
+        if (isset($match[1])) {
+            return $match[1];
```
Comment (framgiaciv3) [inline app/Helpers/helpers.php]: **Reporter**: CHECKSTYLE</br>**Rule**: Generic.Files.LineLength.TooLong</br>**Severity**: warning</br>**File**: app/Helpers/helpers.php</br>**Line**: 15</br>Line exceeds 120 characters; contains 145 characters

File: app/Helpers/helpers.php [hunk 27]
```diff
@@ -2,17 +2,23 @@
...
+        } else {
+            $match[1] = 'https://viblo.asia/announcements/gioi-thieu-state-of-chain-tieu-diem-hackathon-meet-up-blockchain-nam-2018-YWOZrpmp5Q0';
+            return $match[1];
+        }
+    } catch (Exception $e) {
+        return view('error');
     }
```
Comment (framgiaciv3) [inline app/Helpers/helpers.php]: **Reporter**: CHECKSTYLE</br>**Rule**: PSR2.Methods.FunctionClosingBrace.SpacingBeforeClose</br>**Severity**: error</br>**File**: app/Helpers/helpers.php</br>**Line**: 22</br>Function closing brace must go on the next line following the body; found 1 blank lines before brace

**Comment by lxc150896** [comment]
task: https://dev.framgia.com/issues/219943

File: app/Helpers/helpers.php [hunk 28]
```diff
@@ -2,16 +2,21 @@
...
+        $url = 'https://viblo.asia/';
+        $start = config('view.start');
+        $matchLeng = config('view.maxBanner');
+        $content = file_get_contents($url, false, $context, $start, $matchLeng);
+        $data = preg_match('/<div class="container px-0"><img src="(.*?)"/is', $content, $match);
+        if (isset($match[1])) {
+            return $match[1];
+        } else {
+            $match[1] = config('links.banner');
+            return $match[1];
```
Comment (BunnyPi04) [inline app/Helpers/helpers.php:17]: thiếu dòng trắng trc return

File: app/Repositories/Eloquents/OptionEloquentRepository.php [hunk 29]
```diff
@@ -20,6 +20,20 @@ public function getData($data = [], $dataSelect = ['*'])
...
+        $data['textFooters'] = Option::where(['key' => 'text_footer'])->get();
+        $data['banners'] = Option::where(['key' => 'banner'])->get();
+        $data['bannerBooks'] = Option::where(['key' => 'banner_book'])->first();
+        $data['apps'] = Option::where(['key' => 'app'])->get();
+        $data['textBanners'] = Option::where(['key' => 'text_banner'])->get();
+        $data['textApps'] = Option::where(['key' => 'app_text'])->get();
+        $data['contacts'] = Option::where(['key' => 'contact'])->get();
+        $data['address'] = Option::where(['key' => 'address'])->get();
+        $data['emails'] = Option::where(['key' => 'email'])->get();
+        return $data;
```
Comment (BunnyPi04) [inline app/Repositories/Eloquents/OptionEloquentRepository.php:35]: thiếu dòng trắng trc return

File: app/Http/Controllers/SettingController.php [hunk 30]
```diff
@@ -21,24 +21,13 @@ public function getOption($key)
         $conditionText = [
             'key' => $key,
         ];
-
+
```
Comment (BunnyPi04) [inline app/Http/Controllers/SettingController.php]: dư spaces

File: app/Http/ViewComposers/ProfileComposer.php [hunk 31]
```diff
@@ -0,0 +1,35 @@
...
+
+    /**
+     * Create a new profile composer.
+     *
+     * @return  void
+     */
+    protected $option;
+
+    public function __construct(
+        OptionRepository $option
```
Comment (BunnyPi04) [inline app/Http/ViewComposers/ProfileComposer.php]: có mỗi 1 cái thì để nó lên dòng trên đi, cái này đâu phải liệt kê mảng

**Pull request merged**